pkg/syncer/metadatasyncer.go

```diff
@@ -3391,5 +3391,91 @@ func cnsVolumeInfoCRUpdated(oldObj interface{}, newObj interface{}, metadataSync
 	}
 	log.Debugf("cnsVolumeInfoCRUpdated: Old CnsVolumeInfo: +%v, Updated CnsVolumeInfo: +%+v",
 		oldCnsVolumeInfoObj, newCnsVolumeInfoObj)
-	// TODO: add logic to update VolumeSnapshot StoragePolicyUsage.
+
+	// Fetch StoragePolicyUsage instance for storageClass associated with the snapshot.
+	restConfig, err := config.GetConfig()
+	if err != nil {
+		log.Errorf("cnsVolumeInfoCRUpdated: failed to get Kubernetes config. Err: %+v", err)
+		return
+	}
+	cnsOperatorClient, err := k8s.NewClientForGroup(ctx, restConfig, cnsoperatorv1alpha1.GroupName)
+	if err != nil {
+		log.Errorf("cnsVolumeInfoCRUpdated: Failed to create CnsOperator client. Err: %+v", err)
+		return
+	}
+
+	storagePolicyUsageInstanceName := newCnsVolumeInfoObj.Spec.StorageClassName + "-" +
+		storagepolicyv1alpha1.NameSuffixForSnapshot
+	storagePolicyUsageCR := &storagepolicyv1alpha1.StoragePolicyUsage{}
+	err = cnsOperatorClient.Get(ctx, k8stypes.NamespacedName{
+		Namespace: newCnsVolumeInfoObj.Spec.Namespace,
+		Name:      storagePolicyUsageInstanceName},
+		storagePolicyUsageCR)
+	if err != nil {
+		log.Errorf("failed to fetch %s instance with name %q from supervisor namespace %q. Error: %+v",
+			storagepolicyv1alpha1.CRDSingular, storagePolicyUsageInstanceName,
+			newCnsVolumeInfoObj.Spec.Namespace, err)
+		return
+	}
+
+	patchedStoragePolicyUsageCR := storagePolicyUsageCR.DeepCopy()
+	if !newCnsVolumeInfoObj.Spec.ValidAggregatedSnapshotSize {
+		// If ValidAggregatedSnapshotSize field of new CnsVolumeInfo CR is false, then add annotation
+		// "csi.vsphere.missing-snapshot-aggregated-capacity: true" on StoragePolicyUsage CR.
+		log.Infof("Couldn't retrieve aggregated snapshot size for volume %q, adding annotation %s on "+
+			"StoragePolicyUsage CR", common.MissingSnapshotAggregatedCapacity, newCnsVolumeInfoObj.Spec.VolumeID)
+		patchAnnotation := common.MergeMaps(newCnsVolumeInfoObj.Annotations,
+			map[string]string{common.MissingSnapshotAggregatedCapacity: "true"})
+		patchedStoragePolicyUsageCR.Annotations = patchAnnotation
+	} else {
+		if newCnsVolumeInfoObj.Spec.AggregatedSnapshotSize != nil &&
+			storagePolicyUsageCR.Status.ResourceTypeLevelQuotaUsage != nil &&
+			storagePolicyUsageCR.Status.ResourceTypeLevelQuotaUsage.Used != nil {
```

Comment:
What if `storagePolicyUsageCR.Status.ResourceTypeLevelQuotaUsage` is nil? what are the implications? when is this likely to happen?

Reply:
Updated the code to handle cases where storagePolicyUsageCR.Status.ResourceTypeLevelQuotaUsage` is nil